src/client/Client.cc

```diff
@@ -4894,25 +4893,19 @@ void Client::update_snap_trace(const bufferlist& bl, SnapRealm **realm_ret, bool
       ldout(cct, 10) << __func__ << " " << *realm << " seq " << info.seq()
 	       << " <= " << realm->seq << " and same parent, SKIPPING" << dendl;
     }
-        
+
     if (!first_realm)
       first_realm = realm;
     else
       put_snap_realm(realm);
   }
 
-  for (map<SnapRealm*, SnapContext>::iterator q = dirty_realms.begin();
-       q != dirty_realms.end();
-       ++q) {
-    SnapRealm *realm = q->first;
+  for (auto &[realm, snapc] : dirty_realms) {
     // if there are new snaps ?
-    if (has_new_snaps(q->second, realm->get_snap_context())) { 
+    if (has_new_snaps(snapc, realm->get_snap_context())) {
       ldout(cct, 10) << " flushing caps on " << *realm << dendl;
-      xlist<Inode*>::iterator r = realm->inodes_with_caps.begin();
-      while (!r.end()) {
-	Inode *in = *r;
-	++r;
-	queue_cap_snap(in, q->second);
+      for (auto in : realm->inodes_with_caps) {
```

Comment:
```suggestion
      for (auto& in : realm->inodes_with_caps) {
```

`in` may be a smartptr in the future.

Reply:
I have tried the `auto& in` last week, and it will have the following error:

```
/data/ceph/src/client/Client.cc: In member function ‘void Client::update_snap_trace(const bufferlist&, SnapRealm**, bool)’:
/data/ceph/src/client/Client.cc:4908:30: error: cannot bind non-const lvalue reference of type ‘Inode*&’ to an rvalue of type ‘Inode*’
       for (auto &in : realm->inodes_with_caps) {
                              ^~~~~~~~~~~~~~~~
```

> in may be a smartptr in the future.

Yeah, I will add it to my todo list.

Reply:
Try `auto&& in`

Reply:
Done. Thanks.